wasm/lib/src/module.dart

```diff
@@ -454,8 +454,20 @@ class WasmGlobal {
 
   set value(dynamic val) {
     if (_mut == wasmerMutabilityConst) {
-      throw WasmError("Can't set value of const global: $this");
+      throw _WasmErrorModuleImpl("Can't set value of const global: $this");
     }
     runtime.globalSet(_global, _type, val);
   }
 }
+
+class _WasmErrorModuleImpl extends Error implements WasmError {
```

Comment:
This name is a bit odd. Maybe rename to _WasmModuleErrorImpl. Also update the toString method to say WasmModuleError.

Same for the runtime error.

Reply:
> This name is a bit odd

I guess you are referring to the fact that the identifier isn't proper english?

It's actually intentional. Please take a look here https://github.com/modulovalue/extra_pedantic/issues/44#issue-1433890065 where I've briefly described the reasoning behind it.

---

> Also update the toString method to say WasmModuleError. [...] Same for the runtime error.

This is also intentional. The literals returned by the toString implementations only expose the most specific public interface.

`_WasmErrorModuleImpl (WasmError)` -> `'WasmError: $message';`
`_WasmExceptionRuntimeImpl (WasmException)` -> `'WasmException: $message';`
`_WasmErrorRuntimeImpl (WasmError)` -> `'WasmError: $message';`

The reasoning behind this decision is once again [hyrum's law](https://www.hyrumslaw.com/).

If we were to change the literals to e.g. `'WasmErrorModule ...'` then some developer could unintentionally depend on that string belonging to an error that is a `WasmErrorModule`:

```dart
if (e.toString().startsWith('WasmErrorModule')) {
  print("critical error!);
}
```

But since we haven't exposed any public interfaces for errors that are more specific, we shouldn't expose any implementation details through the toString method.

(I anticipate that this package will be used by many people who are primarily more familiar with javascript. It is common to use checks like that in javascript (where it makes sense, because there are no types) so I'm being extra careful here.)

---

Please let me know if you disagree and thank you for the thorough review!

Reply:
> It's actually intentional. Please take a look here [modulovalue/extra_pedantic#44 (comment)](https://github.com/modulovalue/extra_pedantic/issues/44#issue-1433890065) where I've briefly described the reasoning behind it.

I read the reasoning you linked, and I'm not convinced by it. I think the benefit of sticking to the ordinary naming scheme most readers are expecting outweighs the benefit of having a systematic naming scheme. When I read `WasmErrorModule`, I think "ok, this is a module that manages `WasmErrors`".

> The reasoning behind this decision is once again [hyrum's law](https://www.hyrumslaw.com/).
> 
> If we were to change the literals to e.g. `'WasmErrorModule ...'` then some developer could unintentionally depend on that string belonging to an error that is a `WasmErrorModule`:

I'm not really concerned with Hyrum's law. We can't reasonably protect users from all possible mistakes like this. No one should be depending on the stringification of an error. These things change all the time. Just last week we changed the stringification of one of the errors thrown by file system operations in dart:io, which is used by way more people than package:wasm.

Stringifications of errors are intended for debugging, and making these both return WasmError would make them less useful for debugging.